Lecture 4 task: create multilayer perceptron for the selected dataset

In [2]:
# Import necessary libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

heart_path = 'https://raw.githubusercontent.com/DatenFluss/machine-learning-university-course/main/res/heart.csv'
data = pd.read_csv(heart_path)

# Separate the features (X) and the target variable (y)
X = data.drop('output', axis=1)
y = data['output']

# Split the dataset into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Feature scaling: Standardize the features by removing the mean and scaling to unit variance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Create a Multilayer Perceptron (MLP) model using Keras
model = Sequential()

# Add input layer and the first hidden layer
model.add(Dense(16, input_dim=X_train_scaled.shape[1], activation='relu'))  # 16 neurons, ReLU activation

# Add a second hidden layer
model.add(Dense(8, activation='relu'))  # 8 neurons, ReLU activation

# Add the output layer
model.add(Dense(1, activation='sigmoid'))  # 1 neuron (binary classification), Sigmoid activation for probabilities

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
history = model.fit(X_train_scaled, y_train, validation_split=0.2, epochs=50, batch_size=16)

# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(X_test_scaled, y_test)

print(f'Test Accuracy: {test_accuracy:.2f}')

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 3s 28ms/step - accuracy: 0.6200 - loss: 0.6933 - val_accuracy: 0.6735 - val_loss: 0.6122
Epoch 2/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6479 - loss: 0.6249 - val_accuracy: 0.7347 - val_loss: 0.5767
Epoch 3/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6432 - loss: 0.6490 - val_accuracy: 0.7347 - val_loss: 0.5488
Epoch 4/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6938 - loss: 0.5818 - val_accuracy: 0.7347 - val_loss: 0.5266
Epoch 5/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7467 - loss: 0.5416 - val_accuracy: 0.7551 - val_loss: 0.5083
Epoch 6/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7180 - loss: 0.5388 - val_accuracy: 0.7755 - val_loss: 0.4917
Epoch 7/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7485 - loss: 0.5234 - val_accuracy: 0.7551 - val_loss: 0.4771
Epoch 8/50
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7292 - loss: 0.5298 - val_accuracy: 0.7551 - val_loss

Some explanations to the code:

We split the dataset into features (X) and the target (y).
We use StandardScaler: features have zero mean and unit variance.

The MLP model contains an input layer, two hidden layers, and an output layer.
The first hidden layer has 16 neurons, the second hidden layer has 8 neurons, both with ReLU activation.
The output layer uses a single neuron with a sigmoid activation function since it's a binary classification problem.

The model is compiled with the adam optimizer and binary_crossentropy as the loss function for binary classification.
We train the model for 50 epochs with a batch size of 16, using 20% of the training data as a validation set.

Model accuracy is 87%